File: src/workbench/provider-messages.ts
/**
 * Operational facts every provider can report without lending the UI its
 * vocabulary. Drivers translate wire events into this shape; the transcript
 * decides how the fact reads and when its current notice is replaced.
 */
export const PROVIDER_MESSAGE_KINDS = [
  'usage_limit',
  'rate_limit',
  'authentication',
  'authorization',
  'service_unavailable',
  'network',
  'provider_error',
  'retrying',
  'interrupted',
  'model_unavailable',
  'context_limit',
  // A turn that ended without finishing. ACP names five stop reasons and only
  // `end_turn` means the work is done: a refusal the spec asks to be shown,
  // because the prompt it refused is left out of the next one, and the two
  // ceilings (tokens, requests in a turn) cut a turn off mid-work.
  'refusal',
  'turn_limit',
  // The runtime under a local model went away while it was answering. Not
  // `network` — nothing is between the two, they are the same machine — and
  // not `service_unavailable`, which is a provider that answered and said no.
  // This is a process that stopped, and the words for it are different: the
  // reader starts it again (bw-u6cl.9).
  'runtime_stopped',
  'unknown',
] as const;

export type ProviderMessageKind = (typeof PROVIDER_MESSAGE_KINDS)[number];
export type ProviderMessageSeverity = 'info' | 'warning' | 'error' | 'blocking';
export type ProviderMessageScope = 'turn' | 'session' | 'provider' | 'account';

export interface ProviderMessageAction {
  label: string;
  href: string;
}

/**
 * `id` is stable for one condition (for example `usage:session`), not one
 * packet. A later active signal replaces it and a resolved signal clears it.
 */
export interface ProviderMessageSignal {
  id: string;
  kind: ProviderMessageKind;
  phase: 'active' | 'resolved';
  severity: ProviderMessageSeverity;
  scope: ProviderMessageScope;
  /** Provider-supplied explanation retained as detail, never as the headline. */
  detail?: string | null;
  /** ISO instant after which a time-bounded condition is no longer current. */
  retryAt?: string | null;
  /**
   * The provider's own words for when it lifts — `resets 9pm (Asia/Karachi)` —
   * quoted where it named a wall clock rather than an instant.
   *
   * Almost every provider names one this way and none of them names both, so
   * without this the notice carried no time at all: of every usage limit in the
   * owner's own record, not one arrived with a `retryAt` (bw-gao7). It is read,
   * never re-rendered — an hour cannot be recomputed out of a clock face and a
   * zone name without inventing the date it belongs to.
   */
  resets?: string | null;
  action?: ProviderMessageAction | null;
  /**
   * Kept only because records already hold it, and read by nothing.
   *
   * It named an answer-shaped message the notice REPLACED, and both projections
   * acted on it by deleting that message — so a signal filed against the wrong
   * one deleted a real answer, permanently, because a reload is served from the
   * projection. A notice is drawn beside what it is about now, and nothing may
   * mean "delete this message" (bw-by3w).
   */
  sourceMessageId?: string | null;
}

const TITLES: Record<ProviderMessageKind, string> = {
  usage_limit: "You've hit your session limit",
  rate_limit: 'The provider is limiting requests',
  authentication: 'Sign in to continue',
  authorization: 'This action is not allowed',
  service_unavailable: 'The provider is temporarily unavailable',
  network: 'The provider cannot be reached',
  provider_error: 'The provider could not complete this turn',
  retrying: 'Retrying',
  interrupted: 'This turn was interrupted',
  model_unavailable: 'This model is unavailable',
  context_limit: 'This conversation is out of context space',
  refusal: 'The agent declined to continue',
  turn_limit: 'This turn stopped before it finished',
  runtime_stopped: 'The local model runtime stopped',
  unknown: 'The provider reported a problem',
};

function clockReads(iso: string, timeZone?: string): string | null {
  const date = new Date(iso);
  if (!Number.isFinite(date.getTime())) return null;
  return date.toLocaleTimeString([], { hour: 'numeric', minute: '2-digit', ...(timeZone ? { timeZone } : {}) });
}

/**
 * Provider-independent copy used by every live and replayed notice.
 *
 * An instant is preferred and drawn in the reader's own zone. Failing one, the
 * provider's own clause stands as it wrote it: it is the only place the time is
 * written down, and a notice that says a limit was hit without saying when it
 * lifts leaves the reader with nothing to do but guess (bw-gao7).
 */
export function providerMessageReads(signal: ProviderMessageSignal, timeZone?: string): string {
  const reset = signal.retryAt ? clockReads(signal.retryAt, timeZone) : null;
  const zone = reset && timeZone ? ` (${timeZone})` : '';
  const when = reset ? `resets ${reset}${zone}` : signal.resets?.trim() || null;
  return `${TITLES[signal.kind]}${when ? ` · ${when}` : ''}`;
}

export function isProviderMessageKind(value: string): value is ProviderMessageKind {
  return (PROVIDER_MESSAGE_KINDS as readonly string[]).includes(value);
}

export function providerMessageIsCurrent(signal: ProviderMessageSignal, now = Date.now()): boolean {
  if (signal.phase !== 'active') return false;
  if (!signal.retryAt) return true;
  const retryAt = new Date(signal.retryAt).getTime();
  return !Number.isFinite(retryAt) || retryAt > now;
}

const STATUS_LABEL: Record<ProviderMessageKind, string> = {
  usage_limit: 'Limit reached', rate_limit: 'Rate limited', authentication: 'Sign-in required',
  authorization: 'Not allowed', service_unavailable: 'Provider unavailable', network: 'Connection lost',
  provider_error: 'Provider failed', retrying: 'Retrying', interrupted: 'Interrupted',
  model_unavailable: 'Model unavailable', context_limit: 'Context full',
  refusal: 'Declined', turn_limit: 'Stopped short', runtime_stopped: 'Runtime stopped',
  unknown: 'Provider problem',
};

/** The same condition reduced to the compact, colored session-status vocabulary. */
export function providerMessageStatus(signal: ProviderMessageSignal): {
  state: 'stopped' | 'errored' | 'running_tool';
  label: string;
  priority: number;
} {
  const priority = signal.severity === 'blocking' ? 4 : signal.severity === 'error' ? 3 : signal.severity === 'warning' ? 2 : 1;
  return {
    state: signal.severity === 'blocking' ? 'stopped' : signal.severity === 'error' ? 'errored' : 'running_tool',
    label: STATUS_LABEL[signal.kind],
    priority,
  };
}

File: server/src/workbench/provider_messages.rs
//! The conditions this app knows how to draw, and nobody's words for them.
//!
//! What belongs here is provider-neutral by construction: the shape of a
//! condition, the standing it puts a chat in, the word the reader sees. What
//! does not is any reading of a particular kit's prose or of the free-form
//! `data` it hangs off an ACP error — that lives in `kit_words.rs`, which
//! takes a brand, and the test at the foot of this file keeps the two apart
//! (bw-d516).

use serde_json::{json, Value};

/// The same signal, raised from the protocol rather than from prose.
///
/// ACP has a code for this — `AuthRequired`, -32000, "Authentication is
/// required before this operation can be performed" (schema `v1/error.rs`).
/// Every path here read the words instead, so a provider that answered with
/// the code and a terse message came out the far end as "Provider
/// unavailable": indistinguishable from a broken install, with no sign-in
/// offered and no way back into the chat (bw-t26l.20).
pub fn needs_signing_in(detail: &str) -> Value {
    json!({
        "id":"condition:authentication", "kind":"authentication", "phase":"active",
        "severity":"blocking", "scope":"session",
        "detail":detail, "retryAt":Value::Null, "action":Value::Null,
    })
}

/// An ACP error as a sentence a reader can act on.
///
/// A JSON-RPC error is written for a caller: a code, a short message, and a
/// free-form `data` that carries whatever structure the far end wanted to
/// hand over. `Display` for the protocol's error type prints the message and
/// then the whole of `data`, pretty-printed, which is how a chat came to draw
/// `Resource not found: a22f34ef-...: { "uri": "a22f34ef-..." }` in red at the
/// foot of the screen — a sentence, an id nobody can use, and a JSON object,
/// none of it telling the reader what happened or what to do (bw-m15v.3).
///
/// So structure stays out. `data` reaches the reader only when it is itself a
/// sentence, which is what a provider puts there when it has more to say than
/// the message held; an object or a list is for whatever reads structure. And
/// the one condition the protocol names in a way the reader cares about gets
/// the app's own words rather than the wire's, because the wire's are the
/// code's name with an id after them.
pub fn in_plain_words(code: i32, message: &str, data: Option<&Value>) -> String {
    if code == RESOURCE_NOT_FOUND {
        return "The provider could not find something this chat pointed it at.".into();
    }
    let said = message.trim();
    match data
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|more| !more.is_empty())
    {
        Some(more) if said.is_empty() => more.into(),
        Some(more) if !more.contains(said) => format!("{said}: {more}"),
        Some(more) => more.into(),
        None => said.into(),
    }
}

/// ACP's own code for a thing that was asked for and is not there.
pub const RESOURCE_NOT_FOUND: i32 = -32002;

/// The one thing a chat cannot go on without, when it is the thing that is
/// gone. Said where the refusal answered an attempt to take up the chat's own
/// conversation again, rather than anything the conversation mentioned.
pub const NO_SUCH_CONVERSATION: &str =
    "The provider could not find this conversation. It was either cleared or \
     never written down.";

/// The same condition, survived rather than suffered: the chat is carrying on
/// somewhere the provider does not remember, and the reader should know its
/// memory starts here even though the screen still holds everything said.
pub const CONVERSATION_IS_GONE: &str =
    "The provider no longer has this conversation, so it carries on in a fresh \
     one. Everything said above is still here, but the provider cannot read it.";

/// What a chat is doing, in the words of the condition standing over it.
///
/// The server publishes this as a canonical `session.state` event. Live
/// subscribers and restored snapshots consume that same event; the browser
/// never reconstructs state from a notice or restores an earlier activity.
pub fn standing(signal: &Value) -> (&'static str, &'static str) {
    let state = match signal["severity"].as_str() {
        Some("blocking") => "stopped",
        Some("info") | Some("warning") => "running_tool",
        _ => "errored",
    };
    let word = match signal["kind"].as_str().unwrap_or_default() {
        "usage_limit" => "Limit reached",
        "rate_limit" => "Rate limited",
        "authentication" => "Sign-in required",
        "authorization" => "Not allowed",
        "service_unavailable" => "Provider unavailable",
        "network" => "Connection lost",
        "provider_error" => "Provider failed",
        "retrying" => "Retrying",
        "interrupted" => "Interrupted",
        "model_unavailable" => "Model unavailable",
        "context_limit" => "Context full",
        "refusal" => "Declined",
        "turn_limit" => "Stopped short",
        "runtime_stopped" => "Runtime stopped",
        _ => "Provider problem",
    };
    (state, word)
}

/// Which of two conditions standing at once the chat wears.
///
/// The same order the screen sorts by: what stops the chat outranks what only
/// spoiled the turn.
pub fn loudness(signal: &Value) -> u8 {
    match signal["severity"].as_str() {
        Some("blocking") => 4,
        Some("error") => 3,
        Some("warning") => 2,
        _ => 1,
    }
}

pub fn resolved(kind: &str) -> Value {
    json!({"id":format!("condition:{kind}"),"kind":kind,"phase":"resolved","severity":"info","scope":"turn"})
}

#[cfg(test)]
mod tests {
    /// Nothing in here is written in a kit's language.
    ///
    /// The manager, on the fix that first put a vendor's phrasing in this
    /// file: "you need to use proper acp integration, don't put any provider
    /// specific stuff". This is that line, drawn where it can be checked. A
    /// reading that needs a kit's habits needs a brand to go with it, and
    /// `kit_words.rs` is where both live.
    #[test]
    fn native_workbench_services_the_neutral_core_speaks_no_kit_s_language() {
        let source = include_str!("provider_messages.rs");
        let body = source.split("#[cfg(test)]").next().unwrap().to_lowercase();
        for word in [
            "claude",
            "codex",
            "anthropic",
            "openai",
            "gemini",
            "goose",
            "copilot",
            "you've hit",
            "try again at",
            "rate_limit_info",
            "resetsat",
        ] {
            assert!(!body.contains(word), "the neutral core says `{word}`");
        }
    }
}
